### DISCUSSION TITLE
How to subtract music  available in Pulseaudio stream!

### DISCUSSION THREAD

**Discussion by bektorkhan** [dscripka/openWakeWord#132] [Ideas]
Playing music is probably a normal operation. I can't see any way to subtract the music playing from the mic stream.
It is available in as Snapcastclient playback.stream. At least I got rid off all disturbance from OrangePi Zer 2W USB, by voltage booster MT3608 to 8V and a 7805 analog 5V regulator.. back to 5V to the Cmedia USB Audio.
Anyone tried this.. I am an retired engineer. So I don't have the skills or focus anymore!!
/Bektorkhan

**Comment by dscripka** [comment]
What you are referencing, Acoustic Echo Cancellation (AEC), is certainly possible and is common on telephony applications and commercial digital assistant hardware (e.g., Alexa devices, Google Home, etc.).

Unfortunately, AEC works best when combined with specialized hardware, and with off-the-shelf hardware it can be difficult to implement. I have done some initial experiments with the [Speex AEC](https://www.speex.org/docs/api/speex-api-reference/group__SpeexEchoState.html) implementation, and it helps somewhat. I think there is more potential here, but unfortunately I may not have the time in the near future to explore further.

**Comment by bektorkhan** [comment]
Thanks.. but its not the same since we know the majority of the sound
recorded, since we are playing it!!
So with some latency and level correction it would improve the input
considerably.. I'll collect some energy to try something..


Den sön 17 mars 2024 kl 16:03 skrev dscripka ***@***.***>:

**Comment by bektorkhan** [comment]
Thanks.. but its not the same since we know the majority of the sound recorded, since we are playing it!!
So with some latency and level correction it would improve the input considerably.. I'll collect some energy to try something..
